File: 1963-minimum-number-of-swaps-to-make-the-string-balanced/1963-minimum-number-of-swaps-to-make-the-string-balanced.cpp
class Solution {
public:
    bool check(stack<char> &st,char ch){
        if(st.top() == '[' && ch == ']')
            return true;
        return false;
    }
    
    int minSwaps(string s) {
        if(s.size() == 0)
            return 0;
        
        int i = 0;
        stack<char> st;
        while(i<s.size()){
            while(!st.empty() && check(st,s[i])){
                st.pop();
                i++;
            }
            st.push(s[i]);
            i++;
        }
        int open = st.size()/2;
        return (open+1)/2;
        
    }
};


/*
]] [[

]]] [[[


pattern - (no. of opening brackets after removing balanced strings + 1)/2;


*/

File: 1963-minimum-number-of-swaps-to-make-the-string-balanced/1963-minimum-number-of-swaps-to-make-the-string-balanced.java
class Solution {
    public boolean check(Stack<Character> st,char ch){
        if(st.peek() == '[' && ch == ']')
            return true;
        return false;
    }
    
    int minSwaps(String s) {
        Stack<Character> st = new Stack<>();
        int i = 0;
        while(i<s.length()){
            while(!st.isEmpty() && i<s.length() && check(st,s.charAt(i))){
                st.pop();
                i++;
            }
            if(i<s.length()){
                st.push(s.charAt(i));
                i++;
            }
        }
        int open = st.size()/2;
        return (open+1)/2;
    }
};